src/library/scala/collection/immutable/List.scala

```diff
@@ -485,6 +485,51 @@ case object Nil extends List[Nothing] {
  */
 @SerialVersionUID(509929039250432923L) // value computed by serialver for 2.11.2, annotation added in 2.11.4
 final case class ::[B](override val head: B, private[scala] var tl: List[B]) extends List[B] {
+  /** value of 0 means unknown.
+   * Because of the structure of :: and the way the JVM lays out the object
+   * _sizeIfKnown doesn't add to the footprint of the :: object
+   * Its @transient to preserve serialisation format */
+  @transient private var _sizeIfKnown = 0
```

Comment:
It might not be the case on the/a JVM, but it's definitely going to add a field to every `::` in JavaScript, and in Native. That seems a high price to pay to get `knownSize` working on Lists, which is decidedly known not to have a cheap size in the first place.

Please don't do that.

Reply:
@sjrd what is the Native object layout? Most layout schemes these days are 64bit aligned on modern processors, and it would be good (at least for me) to know where the optimisations would not apply

maybe we can add this a platform specific method  but that would require additional tooling

The motivation for this is to improve `length` and `size` not the hints like `sizeHintIfCheap`, they just come for free 

I have a (closed source) implementation of part of this change that we have internally running via instrumentation and bytecode rewriting, which did yield measurable improvements on a variety of applications

Reply:
> what is the Native object layout? Most layout schemes these days are 64bit aligned on modern processors, and it would be good (at least for me) to know where the optimisations would not apply

IIUC, the Native layout is 1 metadata pointer + fields. The most stable GC, immix, allocates everything at a 32-bit boundary. So right now it seems adding a field might not make a difference (since it would go from 24 to 32), however Native doesn't have anywhere to store an identity hash code yet, nor a monitor. I expect we'll eventually need one more word to account for those.

> The motivation for this is to improve length and size not the hints like sizeHintIfCheap, they just come for free

I don't know what your codebase is doing, and you probably have your reasons, but I would expect that using a `List` might not be the best thing to do if you need to use `size` often enough that it matters.

Reply:
> * Because of the structure of :: and the way the JVM lays out the object

Is this JVM specified behaviour or unspecified?  We wouldn't want to suddenly start paying for this.

Reply:
> Implementation details that are not part of the Java Virtual Machine's specification would unnecessarily constrain the creativity of implementors. For example, the memory layout of run-time data areas, the garbage-collection algorithm used, and any internal optimization of the Java Virtual Machine instructions (for example, translating them into machine code) are left to the discretion of the implementor.

https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-2.html

So it seems unwise to do this, IMO.